README.md
<a href="https://bioconductor.org/checkResults/release/bioc-LATEST/MANOR/" style="vertical-align: top;"><img src="https://bioconductor.org/shields/build/release/bioc/MANOR.svg" alt="Bioc release build results"/> (release)</a>
<a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/MANOR/" style="vertical-align: top;"><img src="https://bioconductor.org/shields/build/devel/bioc/MANOR.svg" alt="Bioc devel build results"/> (devel)</a>
<a href="https://ci.appveyor.com/project/pneuvial/MANOR"><img src="https://ci.appveyor.com/api/projects/status/github/pneuvial/MANOR?svg=true" alt="Build status"></a>
<a href="https://github.com/pneuvial/MANOR/actions?query=workflow%3AR-CMD-check"><img src="https://github.com/pneuvial/MANOR/workflows/R-CMD-check/badge.svg" alt="Build status"></a>


# MANOR: CGH Micro-Array NORmalization

MANOR is an [R] package on [Bioconductor] for importation, normalization, visualization, and quality control functions to correct identified sources of variability in array-CGH experiments.


## Links

* MANOR website: http://bioinfo.curie.fr/projects/manor/index.html
* Bioconductor release version: https://www.bioconductor.org/packages/MANOR/
* Bioconductor devel version: https://www.bioconductor.org/packages/devel/bioc/html/MANOR.html


[R]: https://www.r-project.org/
[Bioconductor]: https://www.bioconductor.org/



NEWS.md
# MANOR 1.59.5

 * Merge inst/NEWS to NEWS.md

# MANOR 1.59.4

 * Update broken links in vignette.

 * Update broken links in vignette.

 * Add BugReports link.

 * Add missing dependencies 'rmarkdown' and 'bookdown' to Suggests.
 

# MANOR 1.59.3

 * Package undeprecated by the Bioconductor maintainers.


# MANOR 1.59.2

 * Turned old .Rnw vignette to .Rmd vignette.
 
 * GLAD no more listed in Depends (only in Imports).
 
 * Moved to UTF-8 encoding (to cope with French accents).
 
 * Updated maintainer email address.
 
 * Fixed "register native routines and to disable symbol search".
 
 * Use R's RNG instead of the one from the system in C code.

# MANOR 1.37.2 (2014-10-01)

 * Removed unused C functions for I/O in 'lib_io'.

# MANOR 1.37.1 (2014-10-01)

 * Minor changes to C code to avoid WARNINGS upon R CMD check.
 
 * Changed 'par' settings in vignette to avoid 'Figure margins too large' errors.


# MANOR 1.33.1 (2013-07-03)

 * Replaced calls to 'exit', 'fprintf' which now raise a WARNING upon check.

 * Iteration index not printed anymore in 'nem' (raised an error when compiling the vignette).

# MANOR 1.23.3 (2011-03-18)

 * Updated calls to 'GLAD:::daglad' in the vignette to fix an error
caused by changes in the defaults of 'daglad'.

# MANOR 1.21.1 (2010-10-01)

 * Cleaned 'data/flags.RData" which contained objects from an old 'globalenv'.

 * Updated maintainer's email address.

# MANOR 1.19.1 (2010-01-24)

 * Added (back) 'intensity.flag' to 'data/flags.RData' (had been removed since v 1.12.0 for an unknown reason).

# MANOR 1.15.4 (2009-01-15)

 * misplaced alignment tab in man/spatial.Rd.

# MANOR 1.15.3 (2009-01-13)

 * updated references in .Rd files.

 * fixed warnings due to incorrect use of \item in .Rd files.

# MANOR 1.15.2 (2009-01-06)

 * 'norm' and 'sort' are now S3 methods as well

# MANOR 1.15.1 (2009-01-04)

 * (almost) one file per function in R/

 * removed empty section \details in man/qscore.Rd

 * added a NAMESPACE

 * removed inst/doc/Makefile (not needed anymore because no html output required)

# MANOR 1.15.1 (2009-01-02)

 * removed another non-standard keyword

# MANOR 1.15.1 (2009-01-01)

 * only one keyword per \keyword entry...

# MANOR 1.15.0 (2008-12-31)

 * now use standard "keyword"s

 * changed \link{\code{stuff}} into \code{\link{stuff}}

# MANOR 1.15.0 (2008-11-26)
 
 * filled in "keyword" sections in .Rd files.
 
 * removed empty "examples" sections from .Rd files.
 
 * initialized a few variables upon declaration in C code to prevent warnings in R CMD CHECK.

# MANOR 1.13.2 (2008-09-23)
 
 * modification de la fonction cv pour retourner NA lorsque toutes la valeurs du vecteur sont à NA
 
 * modification de la function getChromosomeArm pour que cytoband ne soit pas positionnée à NULL

# MANOR 1.13.1 (2008-09-04)

 * added a CHANGELOG

 * updated outdated reference in the .bib file

 * changed the definition of flag "rep.flag" to avoid the error now caused by sd(NA, na.rm=TRUE)


R/utils.R
## useful auxiliary functions

## Copyright (C) 2004 Institut Curie
## Author(s): Pierre Neuvial (Institut Curie) 2004
## Contact: manor@curie.fr


### coefficient of variation
cv <- function(x)
{
## 	x <- x[!is.na(x)]
## 	if(length(x) > 0)
## 		ret <- sd(2^x)/mean(2^x)
## 	else
## 		ret <- as.numeric(NA)
## 	ret

### modification de la fonction le 23/09/2008 (P. Hupe)
### permet de retourner NA quand toutes les valeurs de x sont ? NA (sinon il y a une erreur)
  if (length(attr(na.omit(x),"na.action"))==length(x))
    return(NA)
  else 
    return(sd(2^x,na.rm=TRUE)/mean(2^x,na.rm=TRUE))

        
}

### rename variables of a data.frame
rename <- function(data, old.vars, new.vars)
{
    if (length(old.vars)!=length(new.vars))
        warning("old.vars and new.vars have different lengths...\n")
    m <- match(old.vars, names(data))
    w <- which(is.na(m))
    if (length(w))
    {
        for (i in 1:length(w)) warning(paste("Variable not found: ", old.vars[w[i]], "\n", sep=""))
        m <- m[-w]
        new.vars <- new.vars[-w]
    }
    names(data)[m] <- new.vars
    data
}

### aggregate.data.frame (light version)
my.aggregate.data.frame <- function (x, by, FUN, ...)
{
    x <- as.data.frame(x)
    b <- x[, by]
    y <- lapply(x, tapply, list(b), FUN, ..., simplify = FALSE)
    if (any(sapply(unlist(y, recursive = FALSE), length) > 1))
        stop("`FUN' must always return a scalar")
    y <- data.frame(lapply(y, unlist, use.names = FALSE))
    names(y) <- c(names(x))
    y
}

## from character/factor to numeric chromosome index...
chrom <- function(c)
{
    char.c <- as.character(c)
    char.c[(char.c=="X")] <- 23
    char.c[(char.c=="Y")] <- 24
    char.c[(char.c=="Z")] <- 25
    char.c[(char.c=="")] <- 0
    chr <- as.numeric(char.c)
    chr
}

flag.aggregate.arrayCGH <- function(x)
                                        # transfer clone-level information to all spots from a clone
{
    if (length(grep("V", x)))
                                        # bad CLONE (the 3 spots have to be excluded)
        ret <- "V"
    else
    {
        if (length(grep("T", x)))
                                        # bad CLONE (the 3 spots have to be excluded)
            ret <- "T"
        else
            ret <- paste(x, collapse="")
    }
    ret
}

## check that data.names contain test.names
check.names <- function(test.names, data.names, optional=TRUE) {
    w <- which(is.na(match(test.names, data.names)))
    if (length(w)) {
        for (i in 1:length(w)) warning(paste("Variable not found: ", test.names[w[i]], "\n", sep=""))
        if (!optional)
            stop ("Mandatory variables not found")
        test.names <- test.names[-w]
    }
    test.names
}

## add empty lines for design compatibility (pseudo-gpr files)
add.lines <- function(data, sep="_")
{
    df <- NULL
    if (dim(data)[2]!=3)
        stop("Wrong dimension for input data")
    x1 <- data[[1]]
    x2 <- data[[2]]
    x3 <- data[[3]]

    test <- merge(data.frame(x1=1:max(x1)), data.frame(x2=1:max(x2)))
    test <- merge(data.frame(x3=1:max(x3)), test)

    sample <- paste(x1, x2, x3, sep=sep)
    patron <- paste(test$x1, test$x2, test$x3, sep=sep)

##    if(dim(test)[1]!=dim(r)[1])
    if(dim(test)[1]!=dim(data)[1])
    {
        ## design is now compatible with the number of rows in the input file
        w <- which(is.na(match(patron, sample)))              # locate discrepencies
        s <- matrix(unlist(strsplit(patron[w], sep)), ncol=3, byrow=TRUE)
        df <- as.data.frame(s)
        names(df) <- names(data)
    }
    df
}

center <-  function(data, var, by.var) {
    index.name <- "zzz.row.index"
    data[[index.name]] <- 1:(dim(data)[1])
    a <- aggregate(data[[var]], list(b=data[[by.var]]), median, na.rm=TRUE)
    var.median <- paste(var, ".median", sep="")
    names(a)[2] <- var.median
    m <- merge(data, a, by.y="b", by.x=by.var, all.x=TRUE, sort=FALSE)
    m <- m[order(m[[index.name]]),]
    m[[var]]-m[[var.median]]
}

getChromosomeArm <- function(arrayCGH, chrVar="Chromosome", posVar="Position") {
    cytoband <- NULL ## avoids a warning when loading cytoband data...
    data("cytoband", envir = environment()) ### l'appel de la fonction data cr?e l'objet cytoband dans .GlobalEnv
    #cytoband <- get("cytoband", envir=.GlobalEnv) ### on r?cup?re l'objet cytoband contenu dans .GlobalEnv car sinon il reste positionn? ? NULL
    centromere <- cytoband[which(cytoband$Centro==1),]
    rm(cytoband)
    
    chrArmVar <- "ChromosomeArm"
    chrNumVar <- paste(chrVar, "num", sep=".")

    w <- NULL
    for (u in unique(centromere$Chromosome)) {
        wu <- which(centromere$Chromosome==u)
        w <- c(w, wu[which.max(centromere[["Start"]])])
    }

    centromere <- centromere[w,c("Chromosome","Start")]
    centromere$Chromosome <- ChrNumeric(centromere$Chromosome)
    names(centromere) <- c(chrNumVar,"CentroPos")

    arrayCGH$arrayValues[[chrNumVar]] <- ChrNumeric(arrayCGH$arrayValues[[chrVar]])
    arrayCGH$arrayValues <- merge(arrayCGH$arrayValues, centromere, all.x=TRUE)
    arrayCGH$arrayValues[[chrArmVar]] <- "p"
    indNA <- which(is.na(arrayCGH$arrayValues[[chrVar]]))
    if (length(indNA))
        arrayCGH$arrayValues[[chrArmVar]][indNA] <- NA

    ind <- which(arrayCGH$arrayValues[[posVar]]>arrayCGH$arrayValues$CentroPos)
    arrayCGH$arrayValues[[chrArmVar]][ind] <- "q"

    if (length(indNA)<dim(arrayCGH$arrayValues)[1])
        arrayCGH$arrayValues[[chrArmVar]][-indNA] <- paste(arrayCGH$arrayValues[[chrVar]][-indNA], arrayCGH$arrayValues[[chrArmVar]][-indNA], sep="")
    else
        arrayCGH$arrayValues[[chrArmVar]] <- paste(arrayCGH$arrayValues[[chrVar]], arrayCGH$arrayValues[[chrArmVar]], sep="")
    arrayCGH
}
